### PR TITLE
add support for wildcard nats topics

### PR DESCRIPTION
**Pull request by paulwe**

### CODE REVIEW

File: internal/bus/bus_redis.go [hunk 1]
```diff
@@ -73,33 +73,33 @@ func NewRedisMessageBus(rc redis.UniversalClient) MessageBus {
...
-func (r *redisMessageBus) Subscribe(ctx context.Context, channel string, size int) (Reader, error) {
-	return r.subscribe(ctx, channel, size, r.subs, false)
+func (r *redisMessageBus) Subscribe(ctx context.Context, channel Channel, size int) (Reader, error) {
+	return r.subscribe(ctx, channel.Primary, size, r.subs, false)
 }
 
-func (r *redisMessageBus) SubscribeQueue(ctx context.Context, channel string, size int) (Reader, error) {
-	return r.subscribe(ctx, channel, size, r.queues, true)
+func (r *redisMessageBus) SubscribeQueue(ctx context.Context, channel Channel, size int) (Reader, error) {
+	return r.subscribe(ctx, channel.Primary, size, r.queues, true)
```
Comment (boks1971) [inline internal/bus/bus_redis.go:102]: Is there a need to do the migration from `legacy` -> `primary` for redis?

**Review by boks1971 — APPROVED** [review]
LGTM!

File: internal/bus/bus_redis.go (hunk 1 above)
Comment (davidzhao) [reply internal/bus/bus_redis.go:102]: perhaps not.. we could make subject construction specific to each bus

File: pkg/info/channels.go [hunk 2]
```diff
@@ -30,37 +33,109 @@ var channelChar = &unicode.RangeTable{
...
+
+var scratch = &sync.Pool{
+	New: func() any {
+		b := make([]byte, 0, 512)
+		return &b
+	},
+}
+
+func formatClientChannel(service, clientID, channel string) string {
+	p := scratch.Get().(*[]byte)
```
Comment (davidzhao) [inline pkg/info/channels.go:93]: is the type correct here?

File: pkg/server/rpc.go [hunk 3]
```diff
@@ -266,14 +266,17 @@ func (h *rpcHandlerImpl[RequestType, ResponseType]) sendResponse(
 			res.Error = err.Error()
 			res.Code = string(psrpc.Unknown)
 		}
-	} else if response != nil {
+	} else if response.ProtoReflect().IsValid() {
```
Comment (davidzhao) [inline pkg/server/rpc.go:269]: seems this is from a previous PR

**Review by davidzhao — APPROVED** [review]
LG!

File: pkg/info/channels.go (hunk 2 above)
Comment (davidzhao) [reply pkg/info/channels.go:93]: sorry. didn't read this carefully

File: internal/bus/bus.go [hunk 4]
```diff
@@ -24,10 +25,25 @@ const (
...
+	LegacySubCompatiblePub
+	WildcardSubCompatiblePub
+	WildcardSubWildcardPub
+
+	finalChannelMode = WildcardSubWildcardPub
+)
+
+var ChannelMode atomic.Uint32
+
+type Channel struct {
```
[the remark this answers is not in the document]
Comment (paulwe) [reply internal/bus/bus.go:39]: code we push into the bus implementation has to be reimplemented for every bus. the bus implementations used to be like 50 lines each and now redis has this awful monstrosity dedicated to connection reuse and nats has message routing... if there's a compelling reason to change this later we can but for now this seems ok?

File: pkg/info/channels.go (hunk 2 above)
Comment (paulwe) [reply pkg/info/channels.go:93]: using a non-pointer type value in `sync.Pool` makes staticcheck upset `SA6002	Storing non-pointer values in sync.Pool allocates memory`

Comment (boks1971) [reply pkg/info/channels.go:93]: Wow static check catches that. That's great.

File: internal/bus/bus_redis.go (hunk 1 above)
Comment (paulwe) [reply internal/bus/bus_redis.go:102]: we should do it anyway so the implementations are aligned to keep it simple.

File: internal/bus/bus.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (davidzhao) [reply internal/bus/bus.go:39]: yeah no need to change now

**Pull request merged**